### PR TITLE
Storage working group runtime changes. Iteration 2.

### LINKED ISSUE #570: Remove obsolete roles::actors module
Remove obsolete `roles::actors` module because it loses any meaning after the integration with the `bureaucracy` module.

[Root issue is here](https://github.com/Joystream/joystream/issues/544).

### LINKED ISSUE #571: Change the bureaucracy module
### Change the **bureaucracy** module to support the **storage** module integration.

1. Add mint initialization in the genesis.
2. Add method to enumerate workers - it is required to generate a random storage provider.
3. Remove the 'Set storage role parameters' proposal.
4. Remove the 'Evict storage provider' proposal.

[Root issue is here](https://github.com/Joystream/joystream/issues/544).

### LINKED ISSUE #573: Change the runtime crate
### Change the **runtime** crate to support the storage working group integration.

1. Remove or replace `LookupRoles` to query the working group.
2. Data migration:
    - **bureaucracy** mint migration.
    - delete the old data in the data directory module (DataObject).
    - delete the old data in the data object storage registry module 
(DataObjectStorageRelationship).
    - InputContraints values.
3. Add tests.

[Root issue is here](https://github.com/Joystream/joystream/issues/544).

### LINKED ISSUE #630: Restore the old forum code and remove any bureaucracy dependencies.
We should rollback the old forum code and erase all traces of the bureaucracy in the forum crate. We won't use this functionality in the nearest releases.

Subtask:
- Delete `SlashingTerms` artifact.

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
## Changes
- rename "bureaucracy" to the "working group"
- rollback the 'forum-to-working group' integration
- init mint for the working group
- create runtime migrations
    - mint initialization for the storage working group
    - text constraints: `OpeningHumanReadableText`, `WorkerApplicationHumanReadableText`, `WorkerExitRationaleText`
    - remove old data: `data_directory`, `data_object_storage_registry`

### CODE REVIEW

File: node/src/chain_spec.rs [hunk 1]
```diff
@@ -263,10 +267,10 @@ pub fn testnet_genesis(
         }),
         data_object_storage_registry: Some(DataObjectStorageRegistryConfig {
             first_relationship_id: 1,
-        }),
-        actors: Some(ActorsConfig {
-            enable_storage_role: true,
-            request_life_time: 300,
+            storage_working_group_mint_capacity: 0,
```
Comment (mnaamani) [inline node/src/chain_spec.rs:273]: I don't think this is the best place for setting these configuration values.

The way it is done in substrate for the collective module which is instantiated multiple times in the runtime. You can give each instance its own configuration: 
See here https://github.com/paritytech/substrate/blob/1f536e98c509566245056369b752e9a4871de85c/bin/node/cli/src/chain_spec.rs#L284

And the collective module the logic in it's own add_extra_genesis block:
https://github.com/paritytech/substrate/blob/1f536e98c509566245056369b752e9a4871de85c/frame/collective/src/lib.rs#L146

File: runtime-modules/forum/src/lib.rs [hunk 2]
```diff
@@ -9,20 +9,19 @@
...
 use srml_support::{decl_event, decl_module, decl_storage, dispatch, ensure, Parameter};
-use system::{ensure_signed, RawOrigin};
-
-pub use common::constraints::InputValidationLengthConstraint;
-use common::BlockAndTime;
 
 mod mock;
 mod tests;
 
+pub use common::constraints::InputValidationLengthConstraint;
```
Comment (mnaamani) [inline runtime-modules/forum/src/lib.rs]: No need to re export, it is used in `node/src/forum_config/mod.rs` . So It would need be changed there to import from `common` instead.

File: runtime-modules/forum/src/lib.rs [hunk 3]
```diff
@@ -379,14 +382,39 @@ decl_module! {
...
         fn deposit_event() = default;
 
+        /// Set forum sudo.
+        fn set_forum_sudo(origin, new_forum_sudo: Option<T::AccountId>) -> dispatch::Result {
+            ensure_root(origin)?;
+
+            /*
+             * Question: when this routine is called by non sudo or with bad signature, what error is raised?
+             * Update ERror set in spec
+             */
```
Comment (mnaamani) [inline runtime-modules/forum/src/lib.rs:392]:
```
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  DispatchError
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  20
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  0
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  RequireRootOrigin
```

File: runtime-modules/forum/src/lib.rs [hunk 4]
```diff
@@ -379,14 +382,39 @@ decl_module! {
...
+
+            // Hold on to old value
```
Comment (mnaamani) [inline runtime-modules/forum/src/lib.rs:394]: This could be simplified a bit by `take`ing the value so you don't need to kill it later. 
```rust
// Take old value
let old_forum_sudo = <ForumSudo<T>>::take();

// Update forum sudo
if let Some(account_id) = new_forum_sudo.clone() {
    <ForumSudo<T>>::put(account_id);
}
```

File: runtime-modules/service-discovery/src/mock.rs [hunk 5]
```diff
@@ -11,10 +11,10 @@ pub use sr_primitives::{
 
 use srml_support::{impl_outer_event, impl_outer_origin, parameter_types};
 
-mod bureaucracy_mod {
-    pub use bureaucracy::Event;
-    pub use bureaucracy::Instance2;
-    pub use bureaucracy::Trait;
+mod working_group_mod {
+    pub use working_group::Event;
+    pub use working_group::Instance2;
```
Comment (mnaamani) [inline runtime-modules/service-discovery/src/mock.rs]: might make sense to alias `Instance2` to `StorageWorkingGroup` with `as` to make it easier to understand which instance we are working with? In other places as well if it makes sense, this would be important in future as we add more instances. Also why did we skip Instance1 ?

File: runtime-modules/storage/Cargo.toml [hunk 6]
```diff
@@ -73,15 +74,20 @@ default_features = false
 package = 'substrate-membership-module'
 path = '../membership'
 
+[dependencies.minting]
```
Comment (mnaamani) [inline runtime-modules/storage/Cargo.toml]: when removing the add extra genesis block out of the `data_object_storage_registry` module we can remove some unnecessary dependencies from here? Or do we still need all the dependencies of `substrate-working-group-module` as well, ie. minting, stake etc..

File: runtime-modules/storage/src/data_object_storage_registry.rs [hunk 7]
```diff
@@ -121,6 +125,34 @@ decl_storage! {
         /// Keeps a list of storage relationships per content id.
         pub RelationshipsByContentId get(relationships_by_content_id): map T::ContentId => Vec<T::DataObjectStorageRelationshipId>;
     }
+    add_extra_genesis {
```
Comment (mnaamani) [inline runtime-modules/storage/src/data_object_storage_registry.rs]: This can be moved out into the working-group module itself, and the config settings can be set on each instance in the runtime configuration. see my previous comment.

File: runtime-modules/storage/src/tests/data_object_storage_registry.rs [hunk 8]
```diff
@@ -155,3 +157,33 @@ fn test_toggle_ready() {
         );
     });
 }
+
+#[test]
```
Comment (mnaamani) [inline runtime-modules/storage/src/tests/data_object_storage_registry.rs]: I think these tests will also move into the working-group module

File: runtime/src/lib.rs [hunk 9]
```diff
@@ -755,26 +756,11 @@ impl migration::Trait for Runtime {
     type Event = Event;
 }
 
-// Forum bureaucracy
-impl bureaucracy::Trait<bureaucracy::Instance1> for Runtime {
+// Storage working group
+impl working_group::Trait<working_group::Instance2> for Runtime {
```
Comment (mnaamani) [inline runtime/src/lib.rs]: Why the change to Instance2 ? Are you reserving Instance1 for the content working group? Or was instance1 for the forum.. I kind of got confused here. Maybe we can import working_group::Instance2 and working_group::Instance1 with an alias at the top to avoid this confusion and possibly making a mistake references the wrong instance.

File: runtime/src/lib.rs [hunk 10]
```diff
@@ -867,7 +853,6 @@ construct_runtime!(
         Memo: memo::{Module, Call, Storage, Event<T>},
         Members: members::{Module, Call, Storage, Event<T>, Config<T>},
         Forum: forum::{Module, Call, Storage, Event<T>, Config<T>},
-        Actors: actors::{Module, Call, Storage, Event<T>, Config},
```
Comment (mnaamani) [inline runtime/src/lib.rs:870]: Good bye 👍

File: runtime/src/lib.rs [hunk 11]
```diff
@@ -884,9 +869,9 @@ construct_runtime!(
         ProposalsEngine: proposals_engine::{Module, Call, Storage, Event<T>},
         ProposalsDiscussion: proposals_discussion::{Module, Call, Storage, Event<T>},
         ProposalsCodex: proposals_codex::{Module, Call, Storage, Error, Config<T>},
-        // --- Bureaucracy
-        ForumBureaucracy: bureaucracy::<Instance1>::{Module, Call, Storage, Event<T>},
-        StorageBureaucracy: bureaucracy::<Instance2>::{Module, Call, Storage, Event<T>},
+        // --- Working groups
+        // reserved for the future use: ForumWorkingGroup: working_group::<Instance1>::{Module, Call, Storage, Event<T>},
```
Comment (mnaamani) [inline runtime/src/lib.rs:874]: oh hello :) that explains things.

File: runtime-modules/storage/src/data_object_storage_registry.rs [hunk 12]
```diff
@@ -121,6 +125,34 @@ decl_storage! {
...
+        config(worker_exit_rationale_text_constraint): InputValidationLengthConstraint;
+        build(|config: &GenesisConfig<T>| {
+            // Create a mint.
+            let mint_id_result =
+                <minting::Module<T>>::add_mint(config.storage_working_group_mint_capacity, None);
+
+            if let Ok(mint_id) = mint_id_result {
+                <working_group::Mint::<T, working_group::Instance2>>::put(mint_id);
+            } else {
+                print("Failed to create a mint for the storage working group");
```
Comment (mnaamani) [inline runtime-modules/storage/src/data_object_storage_registry.rs]: I would add a panic! here there is no point initializing a chain spec with bad state. And this isn't runtime code so its okay to fail loudly.

File: runtime/src/migration.rs [hunk 13]
```diff
@@ -18,10 +21,24 @@ impl<T: Trait> Module<T> {
         // would be any new storage values that need an initial value which would not
         // have been initialized with config() or build() chainspec construction mechanism.
         // Other tasks like resetting values, migrating values etc.
+
+        Self::initialize_storage_working_group_mint();
+        Self::initialize_storage_working_group_text_constraints();
+
+        // ** Order is important!
+        Self::clear_storage_data_object_registry_data();
+        Self::clear_storage_data_directory_data();
```
Comment (mnaamani) [inline runtime/src/migration.rs]: maybe both steps in a single `fn clear_storage_data()` and avoid need for the "order is important" comment ?

How about also resetting the `next_*_id` state variables to their initial values as well ?

File: runtime-modules/storage/src/data_object_storage_registry.rs [hunk 14]
```diff
@@ -25,10 +25,14 @@
 use codec::{Codec, Decode, Encode};
 use rstd::prelude::*;
 use sr_primitives::traits::{MaybeSerialize, Member, SimpleArithmetic};
-use srml_support::{decl_error, decl_event, decl_module, decl_storage, ensure, Parameter};
+use srml_support::{
+    decl_error, decl_event, decl_module, decl_storage, ensure, print, Parameter, StorageValue,
```
Comment (mnaamani) [inline runtime-modules/storage/src/data_object_storage_registry.rs]: adding `StorageValue` seems un-necessary. Did clippy complain?

**Review by mnaamani — CHANGES_REQUESTED** [review]
Fantastic work.
left a few suggestions.

File: runtime/src/migration.rs (hunk 13 above)
Comment (shamil-gadelshin) [reply runtime/src/migration.rs]: Both fixed.

File: runtime-modules/storage/src/data_object_storage_registry.rs (hunk 12 above)
Comment (shamil-gadelshin) [reply runtime-modules/storage/src/data_object_storage_registry.rs]: Ok.

File: runtime-modules/storage/src/data_object_storage_registry.rs (hunk 14 above)
Comment (shamil-gadelshin) [reply runtime-modules/storage/src/data_object_storage_registry.rs]: Rust thing: a trait declaration should be in scope.

File: runtime/src/lib.rs (hunk 9 above)
Comment (shamil-gadelshin) [reply runtime/src/lib.rs]: Introduced the alias for instances. Instance1 is reserved for the Forum.

File: runtime-modules/service-discovery/src/mock.rs (hunk 5 above)
Comment (shamil-gadelshin) [reply runtime-modules/service-discovery/src/mock.rs]: Introduced the alias for instances. Instance1 is reserved for the Forum.

File: runtime-modules/forum/src/lib.rs (hunk 2 above)
Comment (shamil-gadelshin) [reply runtime-modules/forum/src/lib.rs]: Fixed.

File: runtime-modules/forum/src/lib.rs (hunk 4 above)
Comment (shamil-gadelshin) [reply runtime-modules/forum/src/lib.rs:394]: I didn't change the forum code during this PR - only restored the old version.

File: runtime-modules/forum/src/lib.rs (hunk 3 above)
Comment (shamil-gadelshin) [reply runtime-modules/forum/src/lib.rs:392]: I didn't change the forum code during this PR - only restored the old version. Probably that was old Bedeho's question.

Comment (mnaamani) [reply runtime-modules/forum/src/lib.rs:392]: ok

File: runtime-modules/storage/src/data_object_storage_registry.rs (hunk 14 above)
Comment (mnaamani) [reply runtime-modules/storage/src/data_object_storage_registry.rs]: correct, but if I'm not mistaken the `decl_storage!` macro now brings the necessary ones into scope.. like StorageMap, StorageValue etc.. anyway its not a big deal if it stays.

Comment (shamil-gadelshin) [reply runtime-modules/storage/src/data_object_storage_registry.rs]: Hmm. You are right. It compiles without it. There were errors previously.

File: runtime-modules/storage/Cargo.toml (hunk 6 above)
Comment (shamil-gadelshin) [reply runtime-modules/storage/Cargo.toml]: Moved mint module to the dev-dependencies. All of these modules mint, stake ... etc are required for mocking.

File: runtime-modules/storage/src/data_object_storage_registry.rs (hunk 7 above)
Comment (shamil-gadelshin) [reply runtime-modules/storage/src/data_object_storage_registry.rs]: Done.

File: runtime-modules/storage/src/tests/data_object_storage_registry.rs (hunk 8 above)
Comment (shamil-gadelshin) [reply runtime-modules/storage/src/tests/data_object_storage_registry.rs]: Agree.

File: node/src/chain_spec.rs (hunk 1 above)
Comment (shamil-gadelshin) [reply node/src/chain_spec.rs:273]: Actually, you're right.  I moved the code to the working group and it is much better now. Thanks for the good review.

File: runtime-modules/storage/Cargo.toml (hunk 6 above)
Comment (mnaamani) [reply runtime-modules/storage/Cargo.toml]: oh right of course :+1:

File: runtime-modules/working-group/src/lib.rs [hunk 15]
```diff
@@ -1247,6 +1261,30 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
             )
             .map_err(|e| e.into())
     }
+
+    fn initialize_working_group(
+        opening_human_readable_text_constraint: InputValidationLengthConstraint,
+        worker_application_human_readable_text_constraint: InputValidationLengthConstraint,
+        worker_exit_rationale_text_constraint: InputValidationLengthConstraint,
+        storage_working_group_mint_capacity: minting::BalanceOf<T>,
```
Comment (mnaamani) [inline runtime-modules/working-group/src/lib.rs]:
```suggestion
        working_group_mint_capacity: minting::BalanceOf<T>,
```

File: runtime-modules/working-group/src/lib.rs [hunk 16]
```diff
@@ -1247,6 +1261,30 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
...
+    ) {
+        // Create a mint.
+        let mint_id_result =
+            <minting::Module<T>>::add_mint(storage_working_group_mint_capacity, None);
```
Comment (mnaamani) [inline runtime-modules/working-group/src/lib.rs]:
```suggestion
            <minting::Module<T>>::add_mint(working_group_mint_capacity, None);
```

File: runtime-modules/working-group/src/lib.rs [hunk 17]
```diff
@@ -1247,6 +1261,30 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
...
+
+        if let Ok(mint_id) = mint_id_result {
+            <Mint<T, I>>::put(mint_id);
+        } else {
+            panic!("Failed to create a mint for the storage working group");
```
Comment (mnaamani) [inline runtime-modules/working-group/src/lib.rs]:
```suggestion
            panic!("Failed to create a mint for the working group");
```

Comment (mnaamani) [reply runtime-modules/working-group/src/lib.rs]: I'm not sure if we can identify which instance also in the error message?

File: runtime-modules/working-group/src/tests/mock.rs [hunk 18]
```diff
@@ -124,17 +124,42 @@ impl recurringrewards::Trait for Test {
...
     type Event = TestEvent;
 }
 
 pub type Membership = membership::members::Module<Test>;
-pub type WorkingGroup1 = Module<Test, Instance1>;
+
+pub type TestWorkingGroupInstance = crate::Instance1;
+pub type TestWorkingGroup = Module<Test, TestWorkingGroupInstance>;
+
+pub(crate) const STORAGE_WORKING_GROUP_MINT_CAPACITY: u64 = 40000;
```
Comment (mnaamani) [inline runtime-modules/working-group/src/tests/mock.rs]: `STORAGE_` prefix for these constants here could be removed.. I'm just being picky :)

File: runtime-modules/working-group/src/tests/mod.rs [hunk 19]
```diff
@@ -1592,8 +1617,37 @@ fn set_working_group_mint_capacity_fails_with_mint_not_found() {
...
         let capacity = 15000;
-        let result = WorkingGroup1::set_mint_capacity(RawOrigin::None.into(), capacity);
+        let result = TestWorkingGroup::set_mint_capacity(RawOrigin::None.into(), capacity);
 
         assert_eq!(result, Err(Error::RequireRootOrigin));
     });
 }
+
+#[test]
+fn ensure_setting_genesis_storage_working_group_mint_succeeds() {
```
Comment (mnaamani) [inline runtime-modules/working-group/src/tests/mod.rs]: `_storage` in the names of the test for the working group can be dropped.

File: runtime/src/migration.rs [hunk 20]
```diff
@@ -117,9 +115,10 @@ impl<T: Trait> Module<T> {
 
             potential_id += <T as storage::data_object_storage_registry::Trait>::DataObjectStorageRelationshipId::one();
         }
-    }
 
-    fn clear_storage_data_directory_data() {
+        storage::data_object_storage_registry::NextRelationshipId::<T>::put(<T as storage::data_object_storage_registry::Trait>::DataObjectStorageRelationshipId::one());
```
Comment (mnaamani) [inline runtime/src/migration.rs]: `one()` works because the constant `DEFAULT_FIRST_RELATIONSHIP_ID` defined in the module is equal to 1.
But maybe better to just`kill()` the value, next time it is read it will use the default storage value in this case configured as the constant:

```rust
pub FirstRelationshipId get(first_relationship_id) config(first_relationship_id):
            T::DataObjectStorageRelationshipId = T::DataObjectStorageRelationshipId::from(DEFAULT_FIRST_RELATIONSHIP_ID);
```

**Review by mnaamani — CHANGES_REQUESTED** [review]

File: runtime-modules/proposals/codex/src/proposal_types/mod.rs [hunk 21]
```diff
@@ -47,14 +46,8 @@ pub enum ProposalDetails<MintedBalance, CurrencyBalance, BlockNumber, AccountId,
     /// Balance for the `set content working group mint capacity` proposal
     SetContentWorkingGroupMintCapacity(MintedBalance),
 
-    /// AccountId for the `evict storage provider` proposal
-    EvictStorageProvider(AccountId),
```
Comment (mnaamani) [inline runtime-modules/proposals/codex/src/proposal_types/mod.rs:51]: I just realized we actually need to keep a "dummy" value here to preserve the enumeration index correct when reading state of proposals of this type. 

can we change it to:
`_EvicStorageProvider(AccountId)`

File: runtime-modules/proposals/codex/src/proposal_types/mod.rs [hunk 22]
```diff
@@ -47,14 +46,8 @@ pub enum ProposalDetails<MintedBalance, CurrencyBalance, BlockNumber, AccountId,
...
-
     /// Validator count for the `set validator count` proposal
     SetValidatorCount(u32),
-
-    /// Role parameters for the `set storage role parameters` proposal
-    SetStorageRoleParameters(RoleParameters<CurrencyBalance, BlockNumber>),
```
Comment (mnaamani) [inline runtime-modules/proposals/codex/src/proposal_types/mod.rs:57]: `_SetStorageRoleParameters(RoleParameters<CurrencyBalance, BlockNumber>)`
This means we also need to preserve the `struct RoleParameters` type.

I think we can remove the encapsulated value if we can ensure we don't try to read in the runtime any proposal detail of this variant.

File: runtime-modules/working-group/src/tests/mock.rs (hunk 18 above)
Comment (shamil-gadelshin) [reply runtime-modules/working-group/src/tests/mock.rs]: Fixed.

File: runtime-modules/working-group/src/tests/mod.rs (hunk 19 above)
Comment (shamil-gadelshin) [reply runtime-modules/working-group/src/tests/mod.rs]: Fixed.

File: runtime-modules/proposals/codex/src/proposal_types/mod.rs (hunk 21 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/codex/src/proposal_types/mod.rs:51]: Fixed.

File: runtime-modules/proposals/codex/src/proposal_types/mod.rs (hunk 22 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/codex/src/proposal_types/mod.rs:57]: Fixed.

File: runtime/src/migration.rs (hunk 20 above)
Comment (shamil-gadelshin) [reply runtime/src/migration.rs]: Fixed.

File: runtime-modules/working-group/src/lib.rs (hunk 15 above)
Comment (shamil-gadelshin) [reply runtime-modules/working-group/src/lib.rs]: Fixed.

**Review by mnaamani — APPROVED** [review]
